lgc/state/PalMetadata.cpp

```diff
@@ -1108,26 +1243,47 @@ unsigned PalMetadata::getFragmentShaderBuiltInLoc(unsigned builtin) {
 
 // =====================================================================================================================
 // Returns the calling convention of the first hardware shader stage that will be executed in the pipeline.
-unsigned PalMetadata::getCallingConventionForFirstHardwareShaderStage() {
-  constexpr unsigned hwShaderStageCount = 6;
-  static const std::pair<unsigned, unsigned> shaderTable[hwShaderStageCount] = {
-      {mmSPI_SHADER_PGM_RSRC1_LS, CallingConv::AMDGPU_LS}, {mmSPI_SHADER_PGM_RSRC1_HS, CallingConv::AMDGPU_HS},
-      {mmSPI_SHADER_PGM_RSRC1_ES, CallingConv::AMDGPU_ES}, {mmSPI_SHADER_PGM_RSRC1_GS, CallingConv::AMDGPU_GS},
-      {mmSPI_SHADER_PGM_RSRC1_VS, CallingConv::AMDGPU_VS}, {mmCOMPUTE_PGM_RSRC1, CallingConv::AMDGPU_CS}};
-
-  for (unsigned i = 0; i < hwShaderStageCount; ++i) {
-    auto entry = m_registers.find(m_document->getNode(shaderTable[i].first));
-    if (entry != m_registers.end())
-      return shaderTable[i].second;
+//
+// @param hwStageId :
+unsigned PalMetadata::getCallingConventionForFirstHardwareShaderStage(std::string &hwStageName) {
+  if (m_useRegisterFieldFormat) {
+    auto hardwareStages = m_pipelineNode[Util::Abi::PipelineMetadataKey::HardwareStages].getMap(true);
+    hwStageName = HwStageNames[static_cast<unsigned>(Util::Abi::HardwareStage::Vs)];
+    if (hardwareStages.find(hwStageName) != hardwareStages.end())
+      return CallingConv::AMDGPU_VS;
+
+    hwStageName = HwStageNames[static_cast<unsigned>(Util::Abi::HardwareStage::Hs)];
+    if (hardwareStages.find(hwStageName) != hardwareStages.end())
+      return CallingConv::AMDGPU_HS;
+
+    hwStageName = HwStageNames[static_cast<unsigned>(Util::Abi::HardwareStage::Gs)];
+    if (hardwareStages.find(hwStageName) != hardwareStages.end())
+      return CallingConv::AMDGPU_GS;
+
+    hwStageName = HwStageNames[static_cast<unsigned>(Util::Abi::HardwareStage::Cs)];
+    return CallingConv::AMDGPU_CS;
+  } else {
+    constexpr unsigned hwShaderStageCount = 6;
+    static const std::pair<unsigned, unsigned> shaderTable[hwShaderStageCount] = {
+        {mmSPI_SHADER_PGM_RSRC1_LS, CallingConv::AMDGPU_LS}, {mmSPI_SHADER_PGM_RSRC1_HS, CallingConv::AMDGPU_HS},
+        {mmSPI_SHADER_PGM_RSRC1_ES, CallingConv::AMDGPU_ES}, {mmSPI_SHADER_PGM_RSRC1_GS, CallingConv::AMDGPU_GS},
+        {mmSPI_SHADER_PGM_RSRC1_VS, CallingConv::AMDGPU_VS}, {mmCOMPUTE_PGM_RSRC1, CallingConv::AMDGPU_CS}};
+
+    for (unsigned i = 0; i < hwShaderStageCount; ++i) {
+      auto entry = m_registers.find(m_document->getNode(shaderTable[i].first));
+      if (entry != m_registers.end())
+        return shaderTable[i].second;
+    }
+    return CallingConv::AMDGPU_CS;
```

Comment:
The diff could be simplified by not using if/else. The true case always returns from the function, so there's no need to put the original code into an else block.

Reply:
Got it. Fixed. Make the same modification on `isWave32`.